### PR TITLE
[CostModel] getInstructionCost - improve estimation of costs for length changing shuffles

### PR DESCRIPTION
**Pull request by RKSimon**
Fix gap in the cost estimation for length changing shuffles, by adjusting the shuffle mask and either widening the shuffle inputs or extracting the lower elements of the result

A small step towards moving some of this implementation inside improveShuffleKindFromMask and/or target getShuffleCost handlers (and reduce the diffs in cost estimation depending on whether coming from a ShuffleVectorInst or the raw operands / mask components)

### CODE REVIEW

File: llvm/include/llvm/Analysis/TargetTransformInfoImpl.h [hunk 1]
```diff
@@ -1356,7 +1357,35 @@ class TargetTransformInfoImplCRTPBase : public TargetTransformInfoImplBase {
...
         }
 
-        return CostKind == TTI::TCK_RecipThroughput ? -1 : 1;
+        bool IsUnary = isa<UndefValue>(Operands[1]);
+        NumSubElts = VecSrcTy->getElementCount().getKnownMinValue();
+        SmallVector<int, 16> AdjustMask(Mask.begin(), Mask.end());
+
+        // Widening shuffle - widening the source(s) to the new length
+        // (treated as free - see above), and then perform the adjusted
+        // shuffle at that width.
```
Comment (alexey-bataev) [inline llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:1366]: Dhould not be widening treated as insert subvector? And the add the cost of the insert subvector here?

Comment (RKSimon) [reply llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:1366]: see above

File: llvm/include/llvm/Analysis/TargetTransformInfoImpl.h [hunk 2]
```diff
@@ -1326,28 +1326,29 @@ class TargetTransformInfoImplCRTPBase : public TargetTransformInfoImplBase {
       auto *VecSrcTy = cast<VectorType>(Operands[0]->getType());
       int NumSubElts, SubIndex;
 
+      // TODO: move more of this inside improveShuffleKindFromMask.
       if (Shuffle->changesLength()) {
+        ArrayRef<int> Mask = Shuffle->getShuffleMask();
+
```
Comment (alexey-bataev) [inline llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:1332]: Better to make in a separate NFC patch

File: llvm/include/llvm/Analysis/TargetTransformInfoImpl.h [hunk 3]
```diff
@@ -1356,7 +1357,35 @@ class TargetTransformInfoImplCRTPBase : public TargetTransformInfoImplBase {
...
+            M = M >= NumSubElts ? (M + (Mask.size() - NumSubElts)) : M;
+
+          return TargetTTI->getShuffleCost(
+              IsUnary ? TTI::SK_PermuteSingleSrc : TTI::SK_PermuteTwoSrc, VecTy,
+              AdjustMask, CostKind, 0, nullptr);
+        }
+
+        // Narrowing shuffle - perform shuffle at original wider width and
+        // then extract the lower elements.
+        AdjustMask.append(NumSubElts - Mask.size(), -1);
```
Comment (alexey-bataev) [inline llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:1378]:
```suggestion
        AdjustMask.append(NumSubElts - Mask.size(), PoisonMaskElem);
```

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**